third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -682,6 +686,33 @@ where
         }
     }
 
+    // Check whether the values on the operand stack have the expected return types.
+    // The given types will also be replaced by the expected type, thus if a given
+    // type is more specific (assignable) then expected, it will be replaced by the
+    // type known for the value in the caller context.
+    fn check_and_replace_return_tys(&mut self, current_frame: &mut Frame) -> PartialVMResult<()> {
+        let expected_ret_tys = current_frame.function.return_tys();
+        if expected_ret_tys.is_empty() {
+            return Ok(());
+        }
+        let given_ret_tys = self.operand_stack.last_n_tys_mut(expected_ret_tys.len())?;
+        for (expected, given) in expected_ret_tys.iter().zip(given_ret_tys) {
+            let ty_args = current_frame.function.ty_args();
+            if ty_args.is_empty() {
+                given.paranoid_check_assignable(expected)?;
+                *given = expected.clone()
```

Comment:
So now we clone a type each time we return something? Why do we even want this at runtime? If we return something, and then do a call using this returned type, we will 1) do type susbstitution => reclone this type again, so below lines are wasteful 2) type check